### DISCUSSION TITLE
Item Dropped

### DISCUSSION THREAD

**Discussion by n1c** [saul/demofile#355] [Questions] [answered]
Hello!

I've been poking around and can't find any event for when an item (weapon/equipment) is dropped?

There's a `ItemDrop` user message which doesn't seem to get fired, and `item_remove` event which is really noisy (like all items are removed when a player dies etc).

Not sure if there's something else I'm missing?

Thanks!
Nic.

**Comment by saul** [comment, marked as answer]
Hi @n1c 

Does this help?

```ts
demoFile.entities.on("change", e => {
  const player = (e.entity as unknown) as Player;
  for (let change of e.changes) {
    if (change.tableName === "m_hMyWeapons") {
      if (!change.oldValue) continue;

      const oldValue = change.oldValue as EntityHandle;
      const newValue = change.newValue as EntityHandle;
      const dropped = oldValue.isValid && !newValue.isValid;

      const weapon = (demoFile.entities.getByHandle(
        dropped ? oldValue : newValue
      ) as unknown) as Weapon | null;

      // The weapon was likely deleted
      if (!weapon) continue;

      console.log(
        `${player.name} ${dropped ? "dropped" : "picked up"} ${
          weapon.itemName
        }`
      );
    }
  }
});
```

If you want to ignore weapons dropped on death, then you may need to buffer the dropped items in each tick. Then if a player dies in a given tick, just ignore all of their drops for that tick.

Let me know if you need any more help.

**Reply by n1c** [reply]
I'll experiment and see how it goes; thanks so much!

(such a pity there isn't just a "proper" event for everything 😛 )

**Reply by saul** [reply]
No problem - please update to demofile v2.7.1 first. I've fixed a bug with `EntityHandle.isValid`.
